### DISCUSSION TITLE
Using headless-ui with Fresh

### DISCUSSION THREAD

**Discussion by ClaudiuCeia** [freshframework/fresh#606] [Q&A] [answered]
Hi,

I'm trying to use the Headless UI [`Tabs` component](https://headlessui.com/react/tabs) in a Fresh project, but I'm running into issues. Headless UI seemed like a natural go-to since Fresh ships with tailwind (via twind). It doesn't work out of the box since it's made for React, so I get:

```
This JSX tag's 'children' prop expects a single child of type 'ReactNode | ((bag: TabsRenderPropArg) => ReactElement<any, string | JSXElementConstructor<any>>)', but multiple children were provided.deno-ts(2746)
``` 

Since this is a React library, I tried using the Preact compat layer, like so:

```ts
import { Tabs } from "https://esm.sh/@headlessui/react@1.6.6?alias=react:preact/compat";
```

This got rid of the Typescript errors I had, but I get runtime errors:

```
TypeError: Cannot read properties of undefined (reading '__H')
    at f (https://esm.sh/v87/preact@10.10.0/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/hooks.js:2:217)
    at A (https://esm.sh/v87/preact@10.10.0/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/hooks.js:2:935)
    at R (https://esm.sh/v87/preact@10.10.0/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/hooks.js:2:695)
    at C (https://esm.sh/v87/@headlessui/react@1.6.6/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/react.js:2:6377)
    at https://esm.sh/v87/@headlessui/react@1.6.6/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/react.js:5:66648
    at Object.t (https://esm.sh/v87/preact@10.10.0/X-YS9yZWFjdDpwcmVhY3QvY29tcGF0/deno/compat.js:2:1609)
    at h (https://esm.sh/v86/preact-render-to-string@5.2.0/X-ZC9wcmVhY3RAMTAuOC4y/deno/preact-render-to-string.js:4:1003)
    at h (https://esm.sh/v86/preact-render-to-string@5.2.0/X-ZC9wcmVhY3RAMTAuOC4y/deno/preact-render-to-string.js:9:66)
    at h (https://esm.sh/v86/preact-render-to-string@5.2.0/X-ZC9wcmVhY3RAMTAuOC4y/deno/preact-render-to-string.js:9:66)
    at h (https://esm.sh/v86/preact-render-to-string@5.2.0/X-ZC9wcmVhY3RAMTAuOC4y/deno/preact-render-to-string.js:9:66)
```

Then I check the Preact documentation again and [added all of the rest of the aliases](https://preactjs.com/guide/v10/getting-started#aliasing-react-to-preact):

```ts
import { Tab } from "https://esm.sh/@headlessui/react@1.6.6?alias=react:preact/compat,react-dom/test-utils:preact/test-utils,react-dom:preact-compat,react/jsx-runtime:preact/jsx-runtime";
```

But that brings back the Typescript errors as well as the runtime error (slightly different paths due to different aliases). 

So I'm at a loss here - can I even get this to work?

**Comment by lucacasonato** [comment, marked as answer]
Importing `https://esm.sh/@headlessui/react@1.6.6?alias=react:preact/compat,react-dom:preact/compat,@types/react:preact/compat&deps=preact@10.10.0` works for me.

Ideally `https://esm.sh/@headlessui/react@1.6.6?alias=react:preact/compat,react-dom:preact/compat,@types/react:preact/compat&external=preact` would work, but that seems to be running into some esm.sh bug. cc @ije: the `external=preact` is not applying to most imports it seems.

**Comment by ije** [comment]
instead of complex `?alias`, i added _external all_ pattern in latest updates:

```json
{
  "imports": {
    "react": "http://localhost:8080/preact@10.7.2/compat",
    "preact": "http://localhost:8080/preact@10.7.2",
    "preact/": "http://localhost:8080/preact@10.7.2/",
    "preact-render-to-string": "http://localhost:8080/*preact-render-to-string@5.2.0",
    "swr": "http://localhost:8080/*swr@1.3.0"
  }
}
```

**Comment by OldManFroggy** [comment]
@ClaudiuCeia did you get headless-ui working well, I seem to be having issues with trapping event/state on headless ui controls, for example when toggling the state of a Switch, the switch state seems to break after first assignment. (using doc example from headless-ui site).

**Comment by ClaudiuCeia** [comment]
Thanks @lucacasonato - your suggestion works for me too. I'm not sure what the `external all` pattern is, @ije? I'd love to simplify my import map but the only way I got it working is with:

```json
"@headlessui/react": "https://esm.sh/@headlessui/react@1.6.6?alias=react:preact/compat,react-dom:preact/compat,@types/react:preact/compat&deps=preact@10.10.3"
```

Slightly offtopic - I'd love it if there was a way for me to pin some dependency down to a version and have that reused all over (ie: `preact@10.7.2` in your example) 

---

@scott-ling 

I tried it, seems to work fine for me with:

```ts
const [checked, setChecked] = useState(false);
const toggleChecked = () => setChecked(!checked);
// ... 

<Switch
    checked={checked}
    onChange={toggleChecked}
    class={tw`${
        checked ? "bg-blue-600" : "bg-gray-200"
    } relative inline-flex h-6 w-11 items-center rounded-full`}
    >
    <span class={tw`sr-only`}>Enable notifications</span>
    <span
        class={tw`${
        checked ? "translate-x-6" : "translate-x-1"
        } inline-block h-4 w-4 transform rounded-full bg-white`}
    />
</Switch>
```

**Reply by OldManFroggy** [reply]
Yeah, I must have messed up import since am getting the usual :)

`An error occurred during route handling or page rendering. Error: [object Object] is not a valid HTML tag name in <[object Object] class="bg...`

I will do a fresh project later today, when I get 15min to spare (pun intended..)

**Reply by OldManFroggy** [reply]
oh crud... all I needed to do was a manual refresh of deno cache...  now its working as it should...

now to test all headlessui primitives, before start some serious UI work,

@lucacasonato  @ClaudiuCeia appreciate the assist, reach out anytime.

-S

**Reply by neezer** [reply]
@ClaudiuCeia I think what @ije was saying is that you can omit `?alias=` and instead prefix your dependency with `*` (as in @ije 's example). Couple that with the entry for `"react": "https://esm.sh/preact@10.7.2/compat"` and most of the query string goes away for the following imports.

I tried doing something similar importing [downshift](https://www.downshift-js.com/) in my Fresh project and it worked. Here's what my `import_map.json` looks like now:

```json
{
  "imports": {
    "$fresh/": "https://deno.land/x/fresh@1.0.2/",
    "react": "https://esm.sh/preact@10.10.2/compat",
    "preact": "https://esm.sh/preact@10.10.2",
    "preact/": "https://esm.sh/preact@10.10.2/",
    "preact-render-to-string": "https://esm.sh/*preact-render-to-string@5.2.1",
    "clsx": "https://esm.sh/clsx@1.2.1",
    "downshift": "https://esm.sh/*downshift@6.1.9",
    "compute-scroll-into-view": "https://esm.sh/*compute-scroll-into-view@1.0.17",
    "prop-types": "https://esm.sh/*prop-types@15.8.1",
    "react-is": "https://esm.sh/*react-is@18.2.0",
    "tslib": "https://esm.sh/*tslib@2.4.0"
  }
}
```

Note that you do have to do the same for the dependencies for your packages in `import_map.json` though, so that's a bit of a bummer... otherwise you'll see errors like

> Relative import path "[package-name]" not prefixed with / or ./ or ../ and not in import map from "https://esm.sh/v91/downshift@6.1.9/deno/downshift.external.js"

In my case, `compute-scroll-into-view`, `prop-types`, `react-is`, and `tslib` are all required by `downshift`. Might see if I can write `import_map.json` in [JSON5](https://json5.org) to add some organizational comments or something.

I looked through the esm.sh docs (well, the homepage and a few terse searches through the GH issue list) and couldn't find mention of this feature, so it might be kinda bleeding edge right now. If you've got a link with more details, @ije , that would be handy.

**Reply by ije** [reply]
try `deno run -A https://esm.sh add downshift`, it will resolve the dependencies automatically:

```json
{
  "imports": {
    "downshift": "https://esm.sh/v96/*downshift@6.1.12"
  },
  "scopes": {
    "downshift": {
      "@babel/runtime": "https://esm.sh/v96/@babel/runtime@7.19.4",
      "compute-scroll-into-view": "https://esm.sh/v96/compute-scroll-into-view@1.0.17",
      "prop-types": "https://esm.sh/v96/prop-types@15.8.1",
      "react-is": "https://esm.sh/v96/react-is@17.0.2",
      "tslib": "https://esm.sh/v96/tslib@2.4.0"
    }
  }
}
```

**Reply by alisabzevari** [reply]
This command didn't work for me:
``` bash
$ deno run -A https://esm.sh add downshift   
error: Uncaught (in promise) SyntaxError: Expected double-quoted property name in JSON at position 476
      const { imports, scopes } = JSON.parse(raw);
                                       ^
    at JSON.parse (<anonymous>)
    at loadImportMap (https://esm.sh/:256:40)
    at async Object.add (https://esm.sh/:33:21)
    at async https://esm.sh/:451:5
```

**Reply by ije** [reply]
seems your import maps is not valid JSON

**Comment by TranquilMarmot** [comment]
This doesn't seem to work with all of `@headless-ui/react`.

For example, if you take the example for the menu component and put it into an island:

https://headlessui.com/react/menu

It will fail when you click it with:

```
Uncaught DOMException: Failed to execute 'insertBefore' on 'Node': The node before which the new node is to be inserted is not a child of this node.
```

This seems to be happening in `createRootFragment`:
https://github.com/denoland/fresh/blob/fd4e2bbe17895a3697cd7761a62d95798f6c8d67/src/runtime/main.ts#L4-L25

I figured out that if you wrap `<Menu.Items>` in a `<div>` it works fine. I was able to use render props to set `display: none;` on the `<div>` when the dropdown is closed, so it's not a _huge_ deal but it is ugly and unintuitive.

This seems to be because fresh doesn't like rendering empty fragments in islands. See also this issue: https://github.com/denoland/fresh/issues/502

**Comment by fro-profesional** [comment]
For any one that still having issues wtih some of the components, this import worked for me
```
https://esm.sh/@headlessui/react@1.7.14?alias=react:preact/compat,react-dom:preact/compat,@types/react:preact/compat&external=preact/compat
```

Got it from here https://github.com/denoland/fresh/discussions/432#discussioncomment-3319179
Thanks @czottmann
